x/programs/rust/wasmlanche-sdk/src/memory.rs

```diff
@@ -77,56 +69,89 @@ pub fn split_host_ptr(arg: HostPtr) -> (i64, usize) {
 /// This function is unsafe because it dereferences raw pointers.
 /// # Errors
 /// Returns an [`StateError`] if the bytes cannot be deserialized.
-pub unsafe fn from_host_ptr<V>(ptr: HostPtr) -> Result<V, StateError>
+pub fn from_host_ptr<V>(ptr: i64) -> Result<V, StateError>
 where
     V: BorshDeserialize,
 {
-    let bytes = into_bytes(ptr);
-    from_slice::<V>(&bytes).map_err(|_| StateError::Deserialization)
+    match into_bytes(ptr) {
+        Some(bytes) => from_slice::<V>(&bytes).map_err(|_| StateError::Deserialization),
+        None => Err(StateError::InvalidPointer),
+    }
 }
 
-/// Returns a tuple of the bytes and length of the argument.
+/// Reconstructs the vec from the pointer with the length given by the store
 /// `host_ptr` is encoded using Big Endian as an i64.
 #[must_use]
-pub fn into_bytes(host_ptr: HostPtr) -> Vec<u8> {
-    // grab length from ptrArg
-    let (ptr, len) = split_host_ptr(host_ptr);
-    let value = unsafe { std::slice::from_raw_parts(ptr as *const u8, len) };
-    value.to_vec()
+fn into_bytes(ptr: HostPtr) -> Option<Vec<u8>> {
+    GLOBAL_STORE
+        .with_borrow_mut(|s| s.remove(&(ptr as *const u8)))
+        .map(|len| unsafe { std::vec::Vec::from_raw_parts(ptr as *mut u8, len, len) })
```

Comment:
In your next PR, can you also make sure to leave a `// Safety:` comment on all the `unsafe` blocks?

Reply:
Will do!